[Issue No.24]　Create a registration form and a rules page

groovywave opened the pull request:
# [Issue No.24]　Create a registration form and a rules page
(https://github.com/kenmori/handsonFrontend/blob/master/work/markup/1.md#24)
- 会員登録画面(register.html)を作ります
## [stackblitz](https://stackblitz.com/edit/stackblitz-starters-axeudy?file=README.md,src%2Flesson24%2Findex.html)<br>

仕様
- バリデーションはここではなし
- ユーザー名、メールアドレス、パスワードの入力欄と利用規約に関するチェックボックス(画像参照)がある。
- 送信ボタンがあるが振るまいの実装はしないで良い
- 利用規約のテキストを押すと、モーダルが立ち上がり(前回作ったもので良い)、[ダミーの利用規約](https://terracetech.jp/2021/04/11/gakusyuuyousozairiyoukiyaku/)がテキストとして読める。スクロールが一番下に行ったらチェックボックスはcheckedになる。
- もし開いてもスクロールが下まで行っていなければcheckedはfalseのまま
- checkedがtrueの場合送信ボタンを押下すると別ページのregister-done.htmlに飛ぶ
- register-done.htmlは画面のようなテキストになっている(画面は適当で、遷移できていることが分かれば良い。CSSも書かないでも良い)
<!--
前回実装済
- 画面遷移してから3秒後に解決されるPromiseが返すオブジェクトを元にimgタグを5つつくる。
[以前使ったmy jsonを使う](https://myjson.dit.upm.es/)
- それぞれは.z-indexで重ねた状態。矢印画像をクリックを押すとスライド画像が変わる。
- 5枚中何枚目かを表示して、5/5の場合Nextの矢印はdisabledにする。1/5枚の時はBackボタンはdisabledにする。
-->

今回の対応内容
- 上記の項目。
<!--前回実装済み
-こちらのようなテーブルを画面遷移してから3秒後に解決されるPromiseが返すオブジェクトを元に作り、 idがソートできる機能を作ってください
-ソートが昇順の場合は上矢印がアクティブ、下矢印がdisabled、1,2,3,4,5の順番で表示され、降順の場合はその逆、
-通常時の矢印クリック(クリッカブル領域は2つの矢印です。上下別々のクリッカブル領域でではなく)を押すと画像のように変化します
-->
<!--今回の対応内容
-同じことを年齢でもやってください。
-->

Concerns and areas to focus on<br>
- 関数の配置の仕方（置き所）で悩んでいます。読みやすくするための改善点があればご教示いただきたいです。
<!--実装上、不安なところ、重点的に見てもらいたいところ
- 抽象的にcodingできたとは思いますが、全体的に読みにくいコードにくなってしまいました。-->




<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Introduced a Japanese registration form with validation and modal for terms and conditions.
  - Added a registration completion page with a confirmation message.

- **Documentation**
  - Updated `README.md` with formatting improvements.

- **Style**
  - Implemented new styles for the registration form and completion page using Tailwind CSS.
  - Added `input.css` for additional Tailwind CSS styling configurations.

- **Configuration**
  - Configured Vite and Tailwind CSS in `vite.config.js` and `tailwind.config.js`.
  - Set up a live server port in `.vscode/settings.json`.
  - Added PostCSS configuration for Tailwind CSS integration and Autoprefixer.

- **Refactor**
  - Organized CSS rules in `public/output.css` for better browser consistency and utility classes.

- **Bug Fixes**
  - Corrected file paths and quote consistency in `vite.config.js`.

- **Chores**
  - Added a navigation link to the new lesson in `index.html`.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

@coderabbitai pause

groovywave commented on src/lesson24/js/main.js:
```diff
@@ -0,0 +1,155 @@
+    agreeCheckBox.checked = false;
+  });
+  disableSubmitButton();
+});
```
キャンセルボタンを押したとき、チェックボックスが未チェック、使用不可になります。

groovywave commented on src/lesson24/js/main.js:
```diff
@@ -0,0 +1,151 @@
+      focusableElements[nextFocusedItemIndex].focus();
+    }
+  });
+}
```
tabキーでfocusが移動する範囲をform内に限定しています。

groovywave commented on src/lesson24/js/main.js:
```diff
@@ -0,0 +1,151 @@
+  });
+}
+
+tabAccessControl(document.getElementById("js-form"));
```
tabキーでfocusが移動する範囲をform内に限定しています。

kenmori replied:
Q
こちらはなぜ引数で渡しているのでしょうか？

groovywave commented on src/lesson24/js/main.js:
```diff
@@ -0,0 +1,151 @@
+
+tabAccessControl(document.getElementById("js-form"));
+document.getElementById("js-userName").focus();
+
```
Usernameにfocusしています。

kenmori replied:
読めばわかることを書くというよりは皆が疑問に思うだろうところを説明すると思いますよ

groovywave commented on src/lesson24/js/main.js:
```diff
@@ -0,0 +1,151 @@
+  agreeButton.className =
+    "bg-blue-300 text-white font-bold rounded px-4 py-2 mb-20 hover:bg-blue-700";
+  agreeButton.disabled = false;
+}
```
同意ボタンを有効にします。

groovywave commented on src/lesson24/js/main.js:
```diff
@@ -0,0 +1,151 @@
+    closeModal();
+    enableSubmitButton();
+  });
+}
```
同意ボタンをクリックしたときに、モーダルが閉じ、提出ボタンを利用可能にします。

groovywave commented on src/lesson24/js/main.js:
```diff
@@ -0,0 +1,151 @@
+  submitButton.disabled = false;
+  submitButton.className =
+    "w-full bg-blue-300 hover:bg-blue-700 text-white font-bold py-2 px-4 rounded-lg transition-colors";
+}
```
提出ボタンを利用可能にします。

groovywave commented on src/lesson24/js/main.js:
```diff
@@ -0,0 +1,151 @@
+  submitButton.disabled = true;
+  submitButton.className =
+    "w-full bg-gray-300 text-white font-bold py-2 px-4 rounded-lg transition-colors";
+}
```
提出ボタンを利用不可にします。

groovywave commented on src/lesson24/js/main.js:
```diff
@@ -0,0 +1,149 @@
+    agreeCheckbox.disabled = false;
+    agreeCheckbox.checked = true;
+  });
+}
```
チェックボックスを利用可能にしチェックを入れます。

groovywave commented on src/lesson24/js/main.js:
```diff
@@ -0,0 +1,149 @@
+        "bg-gray-300 text-white font-bold rounded px-4 py-2 mb-20";
+    }
+  });
+}
```
チェックボックスのチェックの有無に応じて同意ボタンを利用可能または不可にします。

groovywave commented on src/lesson24/js/main.js:
```diff
@@ -0,0 +1,149 @@
+      disableSubmitButton();
+    }
+  });
+}
```
チェックボックスのチェックの有無に応じて提出ボタンを利用可能または不可にします。

groovywave commented on src/lesson24/js/main.js:
```diff
@@ -0,0 +1,149 @@
+document.getElementById("js-submitButton").addEventListener("click", (e) => {
+  e.preventDefault();
+  window.location.href = "registration.html";
+});
```
下までスクロールしたときに、同意ボタンが利用可能になります。

groovywave commented on src/lesson24/js/main.js:
```diff
@@ -0,0 +1,149 @@
+  checkboxesToBeChecked();
+  toggleAgreeCheckbox();
+  toggleSubmitCheckbox();
+}
```
下までスクロールしたときに、同意ボタンが利用可能になります。

sae-github commented on assets/css/output.css:
```diff
@@ -0,0 +1,981 @@
+/*
+! tailwindcss v3.4.0 | MIT License | https://tailwindcss.com
+*/
+
+/*
+1. Prevent padding and border from affecting element width. (https://github.com/mozdevs/cssremedy/issues/4)
+2. Allow adding a border to an element by just adding a border-width. (https://github.com/tailwindcss/tailwindcss/pull/116)
+*/
+
+*,
+::before,
```
Q. public内にあるoutput.cssと同じ内容ですか？👀
どちらかでいいと思ったのですが2つある理由が気になりました。

groovywave replied:
I'm sorry ...
I removed that.
77566994cb6d149acda00d4f5d387c19fab83a0a

sae-github commented on src/lesson24/style.css:
```diff
@@ -0,0 +1,54 @@
+  padding-bottom: 20px;
+  position: relative;
+}
+
+.form-control label {
+  color: #777;
+  display: block;
+  margin-bottom: 5px;
+}
+
+.form-control input {
+  border: 2px solid #f0f0f0;
+  border-radius: 4px;
```
Q.このファイルは必要ですか？すべてコメントアウトされているようだったので気になりました！

groovywave replied:
Thank you sae-san:)
I removed that.
77566994cb6d149acda00d4f5d387c19fab83a0a

sae-github commented on src/lesson24/registration.html:
```diff
@@ -0,0 +1,21 @@
+    <meta charset="UTF-8" />
+    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
+    <link href="../input.css" rel="stylesheet" />
+    <!-- <link href="../../dist/output.css" rel="stylesheet" /> -->
```
MUST
不要であれば削除でいいと思います

groovywave replied:
I'm sorry...
a30af77d5c151ed7ce88a1d73e90a63d7fed0ff5

sae-github commented:
MUST
利用規約にチェックが入った状態でモーダルを閉じた際、submitボタンが押せない(disabled)になっているようです。

> スクロールが一番下に行ったらチェックボックスはcheckedになる。
> checkedがtrueの場合送信ボタンを押下すると別ページのregister-done.htmlに飛ぶ

↑が仕様なので、checkedになっていればsubmitできるのが正しいと思いました。


https://github.com/groovywave/JS_Practice/assets/76232904/06dc5add-c83f-437b-bea2-8cce6e13ac76

sae-github commented:
Q. [coderabbita](https://github.com/apps/coderabbitai)はどのように活用していく感じですかね...？

sae-github commented on src/lesson24/index.html:
```diff
@@ -0,0 +1,443 @@
+          id="js-submitButton"
+          type="submit"
+          class="w-full bg-gray-300 text-white font-bold py-2 px-4 rounded-lg transition-colors"
+          disabled="disabled"
+        >
```
IMO
It is a boolean to specify the disabled attribute

> The disabled content attribute is a [boolean attribute](https://html.spec.whatwg.org/multipage/common-microsyntaxes.html#boolean-attribute).

ref : https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#enabling-and-disabling-form-controls:-the-disabled-attribute

groovywave replied:
Thank you sae-san :)
I fixed that.
6d693d2fe2038ddd8c2c1502b1253e2016ebfac6

sae-github commented on src/lesson24/js/main.js:
```diff
@@ -0,0 +1,149 @@
+const username = document.getElementById("username");
+const email = document.getElementById("email");
+const password = document.getElementById("password");
+const password2 = document.getElementById("password2");
+
```
IMO
I feel it would be more readable with a more specific name.
ex. )` rePasswordField` ,` confirmPasswordField`

groovywave replied:
Thank you sae-san :)
I renamed that.
30e5655edd07f56228c786e5482e7fdbbdad03fc

sae-github commented on src/lesson24/js/main.js:
```diff
@@ -0,0 +1,149 @@
+const cancelButton = document.getElementById("js-cancelButton");
+cancelButton.addEventListener("click", () => {
+  closeModal();
+  const agreeCheckBoxes = document.querySelectorAll(
+    '[data-id="js-agreeCheckbox"]'
+  );
```
Q.ここで取得する想定される要素ってなんですかね？
data-idに`js-agreeCheckbox`をしている要素がなさそうです・・・

groovywave replied:
Thank you sae-san :)
I failed to push that.
e8d4040f6cd0d6db6b1bae95a89bc0896b6fdd26

sae-github commented on src/lesson24/js/main.js:
```diff
@@ -0,0 +1,149 @@
+const submitButton = document.getElementById("js-submitButton");
+function enableSubmitButton() {
+  submitButton.disabled = false;
+  submitButton.className =
+    "w-full bg-blue-300 hover:bg-blue-700 text-white font-bold py-2 px-4 rounded-lg transition-colors";
+}
```
IMO
90行目の`disableSubmitButton`内でも同じようなstyleを当てていると思います。
共通するstyle(class)はHTML側で指定するのでもいいと思います。動的に変える必要がないと思うので！

groovywave replied:
This code change state of these buttons, clicking these checkboxes.
https://github.com/groovywave/JS_Practice/assets/76876597/f2eaae5f-735b-4020-88a1-9604777dc585

sae-github replied:
> 共通するstyle(class)

下記は共通していると思います。また動的に変える必要がないスタイルだと思います！
```
text-white font-bold rounded px-4 py-2 mb-20
```

groovywave replied:
I didn’t understand your acvice ;<
I refactored that.
43f0b94deed244b932ad2ad98031ee4bb210727e

sae-github commented on src/lesson24/js/main.js:
```diff
@@ -0,0 +1,149 @@
+  agreeCheckbox.addEventListener("change", () => {
+    agreeButton.disabled = !agreeCheckbox.checked;
+    if (agreeCheckbox.checked) {
+      agreeButton.className =
+        "bg-blue-300 text-white font-bold rounded px-4 py-2 mb-20 hover:bg-blue-700";
+    } else {
+      agreeButton.className =
+        "bg-gray-300 text-white font-bold rounded px-4 py-2 mb-20";
+    }
```
IMO
上記のレビュー同様、こちらも動的に変える必要のないclassはHTML側で指定するといいと思います

groovywave replied:
Thank you sae-san :)
43f0b94deed244b932ad2ad98031ee4bb210727e

sae-github commented on src/lesson24/js/main.js:
```diff
@@ -0,0 +1,149 @@
+  toggleSubmitCheckbox();
+}
+
+document.getElementById("js-submitButton").addEventListener("click", (e) => {
```
IMO
Variables already defined can be used
```suggestion
submitButton.addEventListener("click", (e) => {
```

groovywave replied:
Thank you sae-san :)
I removed the duplicate code.
f4126aa7dd441257102f96f724e07819c513dab4

sae-github commented on src/lesson24/js/main.js:
```diff
@@ -0,0 +1,149 @@
+});
+
+const options = {
+  root: document.getElementById("js-modal"),
```
IMO
Variables already defined can be used
```suggestion
  root: modal,
```

groovywave replied:
Thank you sae-san :)
242d4e929025dcf8178c8fe3f9a2d9b5cf2b1df1

sae-github commented on src/lesson24/js/main.js:
```diff
@@ -0,0 +1,149 @@
+};
+
+const observer = new IntersectionObserver(changeAgreeToClickable, options);
+observer.observe(document.getElementById("js-agreeButton"));
```
IMO
Variables already defined can be used
```suggestion
observer.observe(agreeButton);
```

groovywave replied:
I'm sorry...
I fixed that.
1da7e66e59cd98175c7f8d5a10183551ce55ca21

haru-programming commented:
MUST
課題24はPlaywrightのテストを添付するのが必須だと思います !

haru-programming reviewed (commented):
MUST
課題24はPlaywrightのテストを添付するのが必須だと思います
（塾のルール確認したら17からになってました...！個人的には24からは必須であると助かります🙏）

haru-programming commented:
IMO
モーダル内の同意ボタンの横にチェックボックスもあるのがわかりにくいと思いました。
最後には同意ボタンを押す（これがボタンになっている）と思うので、チェックボックスはなくても良いと思います。
![Screenshot by Dropbox Capture](https://github.com/groovywave/JS_Practice/assets/67426438/869d30f0-aa49-493c-b7f2-66a5f471e439)

haru-programming commented:
IMO
inputがdisabledの時は、cursor: not-allowedだと押せないことがもっとわかりやすくなると思いました !
![Screenshot by Dropbox Capture](https://github.com/groovywave/JS_Practice/assets/67426438/1f3883d9-1a93-41fa-aeda-e765e086fa4a)

groovywave commented:
> cursor: not-allowed

Thank you haru-san:）
I added the class.
7111abcbe3c433cb23879b81d621929e25b2bf1f

https://github.com/groovywave/JS_Practice/assets/76876597/dd5e0987-eb6b-47e3-afb2-4bc6a79ba2bc

groovywave commented:
> IMO モーダル内の同意ボタンの横にチェックボックスもあるのがわかりにくいと思いました。 最後には同意ボタンを押す（これがボタンになっている）と思うので、チェックボックスはなくても良いと思います。 ![Screenshot by Dropbox Capture](https://private-user-images.githubusercontent.com/67426438/293619974-869d30f0-aa49-493c-b7f2-66a5f471e439.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDQxOTMyNDgsIm5iZiI6MTcwNDE5Mjk0OCwicGF0aCI6Ii82NzQyNjQzOC8yOTM2MTk5NzQtODY5ZDMwZjAtYWE0OS00OTNjLWI3ZjItNjZhNWY0NzFlNDM5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTAyVDEwNTU0OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWZhMDg0YjMwNjNmY2FiYzkyMzY4M2YzODA2NTU5OGFjOGEzODVhMTA3ZGYyYjczM2M3ZWE4MTMyNDQ3YzE5MjYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.B7IcVgY7aW8wu0F_uv4sr6O-rn6-PtFDTnsBCEi-eLA)

Thank you haru-san :)
I removed the checkbox.
ae11026d6f83cb27fec8a5b5306bf01e5185f09a

groovywave commented:
> MUST 課題24はPlaywrightのテストを添付するのが必須だと思います （塾のルール確認したら17からになってました...！個人的には24からは必須であると助かります🙏）

Thank you haru-san:)
I checked my code on playwright.
96a98ba34c93d20cc2c3713cca6a24a717784029

sae-github commented:
私のコメントに上書きしているようです...
コメントを新たに追加するようにお願いいたします🙇🏻‍♀️

https://github.com/groovywave/JS_Practice/pull/63#issuecomment-1873142819

![スクリーンショット 2024-01-03 6 38 43](https://github.com/groovywave/JS_Practice/assets/76232904/d8dcfb3e-54f1-434b-8c0d-b0b564e98f97)

sae-github commented:
> I checked my code on playwright.
https://github.com/groovywave/JS_Practice/commit/96a98ba34c93d20cc2c3713cca6a24a717784029
https://github.com/groovywave/JS_Practice/commit/2bf2e62b64e81adfd2a54f67a73fcf4560819667

次回からでいいので、こういった仕様漏れなどのケースでコミットが増える場合は
別のPRを立てるといいと思います！レビュアーの負荷軽減と、mergeまでの期間を短くすることでモチベーションにもいいと思います。

sae-github commented on src/lesson24/js/main.js:
```diff
@@ -0,0 +1,133 @@
+}
+function disableSubmitButton() {
+  submitButton.disabled = true;
+  submitButton.classList.remove("bg-blue-300", "hover:bg-blue-700");
+  submitButton.classList.add("bg-gray-300");
```
IMO
これも動的に付け替える必要はないと思います。class名にで`:disable`をつけて指定することができます
https://tailwindcss.com/docs/hover-focus-and-other-states#form-states
```
 class="disabled:bg-gray-300 disabled:hover:bg-none bg-blue-300 hover:bg-blue-700 
```

haru-programming replied:
Tailwind途中で脱落してしまったので、とても勉強になりました🙏

groovywave replied:
Thank you sae-san :）
I learned a lot of things about tailwindcss!!
6a0a422ad025d8d8d1d6abafc417c0b492dd9780

sae-github commented on src/lesson24/js/main.js:
```diff
@@ -0,0 +1,133 @@
+}
+function disableAgreeButton() {
+  agreeButton.disabled = true;
+  agreeButton.classList.remove("bg-blue-300");
+  agreeButton.classList.add("bg-gray-300");
+}
```
こちらにも下記の内容が適用できますね！
https://github.com/groovywave/JS_Practice/pull/63/files/2bf2e62b64e81adfd2a54f67a73fcf4560819667#r1439902683

groovywave replied:
Thank you sae-san :)
https://github.com/groovywave/JS_Practice/commit/6a0a422ad025d8d8d1d6abafc417c0b492dd9780

sae-github commented on src/lesson24/js/main.js:
```diff
@@ -0,0 +1,133 @@
+function checkboxToBeChecked() {
+  submitCheckbox.disabled = false;
+  submitCheckbox.checked = true;
+  submitCheckbox.classList.remove("cursor-not-allowed");
+}
```
https://github.com/groovywave/JS_Practice/pull/63/files#r1439905054

groovywave replied:
Thank you sae-san :)
I added the class.
![image](https://github.com/groovywave/JS_Practice/assets/76876597/9a2e1396-b615-43f1-8427-8cf94ff40ebf)

274f638a9acbe81c4a9dbe222bdfce9b5b83fe3b

haru-programming commented:
@groovywave 
今のままだとテストに通っているかがレビュワーに伝わらないので、
GitHub Actionsを使用してPR画面でそちらを示せるようにしましょう！
詳細、Discordに載せます！

groovywave commented on tests/JS_lesson.spec.js:
```diff
@@ -0,0 +1,74 @@
+  const takeScreenshot = async (name) => {
+    await page.screenshot({ path: `screenshot/${name}.png` });
+  };
+  await page.goto(
+    "https://stackblitz.com/edit/stackblitz-starters-axeudy?file=README.md"
+  );
+  await page.goto("https://stackblitz.com/edit/stackblitz-starters-axeudy");
+  await page.goto(
+    "https://stackblitz.com/edit/stackblitz-starters-axeudy?file=README.md"
+  );
```
Thank you sae-san:)
I removed the elements that aren't used anywhere.
I don't have particular reason about that.
.(I thought the code was useful for any code.) 

4b23e2e16336cb5c567f329663e1e850b157464d

groovywave commented on tests/JS_lesson.spec.js:
```diff
@@ -0,0 +1,74 @@
+  await page.goto(
+    "https://stackblitz.com/edit/stackblitz-starters-axeudy?file=README.md"
+  );
+  await page
+    .frameLocator('iframe[title="Preview page"]')
+    .getByRole("link", { name: "lesson24" })
+    .click();
```
stackblitzでlesson24のpegeを開くテストのようです。
https://github.com/groovywave/JS_Practice/assets/76876597/bb0affbd-f23c-47bd-a94c-c09647d93549

groovywave commented on src/lesson24/js/main.js:
```diff
@@ -0,0 +1,133 @@
+
+function tabAccessControl(target) {
+  const focusableElementsSelector =
+    'a[href], area[href], input:not([disabled]), select:not([disabled]), textarea:not([disabled]), button:not([disabled]), iframe, object, [tabindex="0"], [contenteditable]';
```
Thank you sae-san:)
I removed the elements that aren't used anywhere.
I don't have particular reason about that.
.(I thought the code was useful for any code.)

https://github.com/groovywave/JS_Practice/commit/4b23e2e16336cb5c567f329663e1e850b157464d

groovywave commented:
> MUST 利用規約にチェックが入った状態でモーダルを閉じた際、submitボタンが押せない(disabled)になっているようです。
> 
> > スクロールが一番下に行ったらチェックボックスはcheckedになる。
> > checkedがtrueの場合送信ボタンを押下すると別ページのregister-done.htmlに飛ぶ
> 
> ↑が仕様なので、checkedになっていればsubmitできるのが正しいと思いました。
> 
>  2024-01-01.13.07.41.mov

Thank you sae-san :）
I fixed addEventlistener so that you can click the button when the checkbox is checked.
(can't click that when the checkbox isn't checked.)
012cfa55d557f80fe54eec0c4078525954b1289d



https://github.com/groovywave/JS_Practice/assets/76876597/24eacc62-3110-4e12-b4e1-69bd6ac56721

groovywave commented:
> > I checked my code on playwright.
> > [96a98ba](https://github.com/groovywave/JS_Practice/commit/96a98ba34c93d20cc2c3713cca6a24a717784029)
> > [2bf2e62](https://github.com/groovywave/JS_Practice/commit/2bf2e62b64e81adfd2a54f67a73fcf4560819667)
> 
> 次回からでいいので、こういった仕様漏れなどのケースでコミットが増える場合は 別のPRを立てるといいと思います！レビュアーの負荷軽減と、mergeまでの期間を短くすることでモチベーションにもいいと思います。

Thank you for your advice :)

groovywave commented:
> nits モーダル内の同意ボタンが薄い背景からhoverで暗くなるのが違和感でした。 薄い状態がdisabledを感じるからかもしれません。 好みの問題かもしれないのでスルーしていただいてもOKです！

Thank you haru-san :)
I changed the class.
9483ace96273cde8a5e6668e7dd4a034b89ba17f

sae-github commented on tests/JS_lesson.spec.js:
```diff
@@ -0,0 +1,72 @@
+  await page.goto("localhost:3000/lesson24/index.html");
+  // await page.goto("http://127.0.0.1:3000/lesson24/index.html");
+  // await page.getByRole("link", { name: "lesson24" }).click();
+  await page.getByPlaceholder("Enter username").click();
+  await page.getByPlaceholder("Enter username").press("Tab");
+  await page.getByPlaceholder("Enter email").press("Tab");
```
Q. ここでテストしたいことって何ですかね？
tabで移動した際に対象の要素にforcusが当たっているかがテスト項目になるのではないかと思いました。
すみません、私もテストがまだ苦手なので違ったら教えてください・・
https://playwright.dev/docs/input#focus-element

groovywave replied:
tests using focus() and click() has failed ;<
the test using focus() and press("Enter") has passed but transition has failed ;<

sae-github replied:
> tests using focus() and click() has failed ;<
the test using focus() and press("Enter") has passed but transition has failed ;<

Q. すみません、上記のコメントが理解できなかったので日本語で補足いただけると嬉しいです。
> but transition has failed 

transitionとは何を指してますか？

groovywave replied:
伝えるのが下手で申し訳ありません。
ページ遷移のつもりでtransitionと書いてしまいました。

”利用規約”を対象にfocus() と press("Enter") の実行テストは通ったのですが、screenshotからページ遷移は確認できませんでした。

haru-programming replied:
@groovywave 
さえさんのおっしゃる通り、Tab遷移して→focusが当たったことを確認するテストを書くべきだと思いました👀

別のところの例ですが、
```
    await page.keyboard.press('Tab');
    await expect(page.getByRole('button', { name: 'パスワードが表示されます' })).toBeFocused();
```
expectと, toBeFocused()を使うと実現できると思います !

groovywave replied:
Thank you haru-san sae-san :)
I wrote the code.
57da8bc226bb1a837e848878c0ad850073259c69

sae-github commented on tests-examples/demo-todo-app.spec.js:
```diff
@@ -0,0 +1,449 @@
+// @ts-check
+const { test, expect } = require('@playwright/test');
+
+test.beforeEach(async ({ page }) => {
+  await page.goto('https://demo.playwright.dev/todomvc');
+});
+
+const TODO_ITEMS = [
+  'buy some cheese',
+  'feed the cat',
```
MUST
不要なファイルだと思います

groovywave replied:
Thank you  sae-san :)
I removed the file.

sae-github commented on tests/JS_lesson.spec.js:
```diff
@@ -0,0 +1,72 @@
+  await page.getByPlaceholder("Enter password", { exact: true }).press("Tab");
+  await page.getByPlaceholder("Enter password again").press("Tab");
+  await page.getByRole("link", { name: "利用規約" }).press("Enter");
+  await page.getByLabel("利用規約", { exact: true }).press("ArrowDown");
+  await page.getByLabel("利用規約", { exact: true }).press("ArrowDown");
+  await page.getByLabel("利用規約", { exact: true }).press("ArrowDown");
```
Q. await page.getByLabel("利用規約", { exact: true }).press("ArrowDown");が複数あるのには理由があるのですか？

groovywave replied:
playwright can't recognize mouse scroll so I use ArrowDown.

https://github.com/groovywave/JS_Practice/assets/76876597/11bc5c35-e3c5-42ab-a124-c590a7f30fd5

groovywave replied:
tests using focus() and click() has failed ;<
the test using focus() and press("Enter") has passed but transition has failed ;<
![image](https://github.com/groovywave/JS_Practice/assets/76876597/97bb9640-75e7-463f-8f8c-44578163eb88)

sae-github replied:
IMO
知見がなくて申し訳ないのですが、下記のZennの記事のようにすると同じ記述は避けられそうです！
ref: https://zenn.dev/takky94/articles/43291a50292c6a

対象の値を探すまでスクロールするような処理だと思います。(ちょっと自分も調べきれてないですが)
```js
  let target = await page.$$('text=利用規約を読み、同意しました')
  while (!target.length) {
    await page.mouse.wheel(0, 10000) 
    await page.getByText("利用規約を読み、同意しました", { exact: true }).click();
    await takeScreenshot("click agreeButton");
  }
```

groovywave replied:
Thank you sae-san :）
Your code can work!
I learned how to code playwright :)
![image](https://github.com/groovywave/JS_Practice/assets/76876597/1cdcd8e4-c26e-49b9-8a83-40b607ddbcb5)

haru-programming replied:
@groovywave @sae-github 
私もテストまだまだ初心者なのですが、該当箇所このように書いてました ! 
```js
    await page.getByTestId('modal-contents').click();
    await page.getByTestId('last-sentence').scrollIntoViewIfNeeded(); // こちら
    await page.getByRole('button', { name: 'Close' }).click();
```

[scrollIntoViewNeeded()](https://playwright.dev/docs/api/class-locator#locator-scroll-into-view-if-needed)
> [IntersectionObserver](https://developer.mozilla.org/en-US/docs/Web/API/Intersection_Observer_API)の定義に従って完全に表示されていない限り、要素をビューにスクロールしようとします

最初はさえさんの方法で実装していました !
最適解が分かりましたら教えてください 🙏

sae-github replied:
わーーありがとうございます🥹🥹勉強になる。。(getByTestIdも初めて知った)
今回のケースは最下部のコンテンツが決まっているので、はるさんの案が良さそう。
完全に動的なコンテンツ(無限スクロールとか)で最後の要素が決まってなければループで回すとかの方がパフォーマンス的にも良さそうですね！

groovywave replied:
Thank you haru-san sae-san :)
I added .scrollIntoViewIfNeeded() in my code.
57da8bc226bb1a837e848878c0ad850073259c69

target element has disabled, so I also added .press("ArrowDown") .
It's tricky but the code can't work without this.

sae-github reviewed (commented):
一旦気になった箇所コメント入れました！

groovywave commented on tests/JS_lesson.spec.js:
```diff
@@ -0,0 +1,72 @@
+// @ts-check
+
```
Thank you sae-san :)
I named the 3 tests.
25326cd3184b721998495af2ce7188d693d810b5

haru-programming reviewed (commented):
@groovywave 
> Thank you haru-san :)
I changed the class and adjusted the color of buttons when hovering.
https://github.com/groovywave/JS_Practice/commit/9483ace96273cde8a5e6668e7dd4a034b89ba17f

■Comment
hoverした時に色が薄くなる方が自然かなと思いコメントしていたので修正が確認できなかったのですが、
全然nitsなので修正は大丈夫です🙇‍♀️

■MUST
コミット履歴にあった該当箇所のtailwindに不要な空白が多くあるようなので、こちらは直した方が良いと感じました。

■Q
また、上記のコミットに一緒にあったこちらは必要でしょうか？🙇‍♀️
![Screenshot by Dropbox Capture](https://github.com/groovywave/JS_Practice/assets/67426438/47c2cfcb-6bb0-464f-a14d-b8bf255ecdf1)

haru-programming commented on src/lesson24/index.html:
```diff
@@ -0,0 +1,441 @@
+        </div>
+        <div class="flex-auto items-center text-center justify-center">
+          <button
+            data-modal-toggle="popup-modal"
+            type="button"
+            id="js-agreeButton"
+            class= "bg-blue-300  text-white font-bold rounded px-4 py-2 mb-20
+            hover:bg-blue-400
+            disabled:bg-gray-300 "
+            data-id="js-agreeButton"
+            disabled
+            data-id="focusable"
```
MUST
data-idが2つあるようです👀使用されていますか？

Q
data-modal-toggle="popup-modal"も使用されていますか？

groovywave replied:
I added these data-ids but didn't use them after all.

groovywave commented:
> @groovywave
> 
> > Thank you haru-san :)
> > I changed the class and adjusted the color of buttons when hovering.
> > [9483ace](https://github.com/groovywave/JS_Practice/commit/9483ace96273cde8a5e6668e7dd4a034b89ba17f)
> 
> ■Comment hoverした時に色が薄くなる方が自然かなと思いコメントしていたので修正が確認できなかったのですが、 全然nitsなので修正は大丈夫です🙇‍♀️
> 
> ■MUST コミット履歴にあった該当箇所のtailwindに不要な空白が多くあるようなので、こちらは直した方が良いと感じました。
> 
> ■Q また、上記のコミットに一緒にあったこちらは必要でしょうか？🙇‍♀️ ![Screenshot by Dropbox Capture](https://private-user-images.githubusercontent.com/67426438/295209746-47c2cfcb-6bb0-464f-a14d-b8bf255ecdf1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDQ4NTc4OTgsIm5iZiI6MTcwNDg1NzU5OCwicGF0aCI6Ii82NzQyNjQzOC8yOTUyMDk3NDYtNDdjMmNmY2ItNmJiMC00NjRmLWExNGQtYjhiZjI1NWVjZGYxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMTAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTEwVDAzMzMxOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWUyZjNiZDQwYTkxZTc5NjhmNjk3NzI2YTRjNDU4ZWEzOTY0YTA5MTE4NWIyYWNmZjJiOTQyNjMyZTdlZTMxZmImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.V-J5WJc3DOWQ3PVqmblwI_qzYpV4_Elq2XNii9jgkaw)

Thank you haru-san :)
I installed prettier-plugin-tailwindcss
eaf69bcf87af238e5a61ed71e80207912cf0e133

groovywave commented:
> > @groovywave
> > > Thank you haru-san :)
> > > I changed the class and adjusted the color of buttons when hovering.
> > > [9483ace](https://github.com/groovywave/JS_Practice/commit/9483ace96273cde8a5e6668e7dd4a034b89ba17f)
> > 
> > 
> > ■Comment hoverした時に色が薄くなる方が自然かなと思いコメントしていたので修正が確認できなかったのですが、 全然nitsなので修正は大丈夫です🙇‍♀️
> > ■MUST コミット履歴にあった該当箇所のtailwindに不要な空白が多くあるようなので、こちらは直した方が良いと感じました。
> > ■Q また、上記のコミットに一緒にあったこちらは必要でしょうか？🙇‍♀️ ![Screenshot by Dropbox Capture](https://private-user-images.githubusercontent.com/67426438/295209746-47c2cfcb-6bb0-464f-a14d-b8bf255ecdf1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDQ4NTc4OTgsIm5iZiI6MTcwNDg1NzU5OCwicGF0aCI6Ii82NzQyNjQzOC8yOTUyMDk3NDYtNDdjMmNmY2ItNmJiMC00NjRmLWExNGQtYjhiZjI1NWVjZGYxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMTAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTEwVDAzMzMxOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWUyZjNiZDQwYTkxZTc5NjhmNjk3NzI2YTRjNDU4ZWEzOTY0YTA5MTE4NWIyYWNmZjJiOTQyNjMyZTdlZTMxZmImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.V-J5WJc3DOWQ3PVqmblwI_qzYpV4_Elq2XNii9jgkaw)
> 
> Thank you haru-san :) I installed prettier-plugin-tailwindcss [eaf69bc](https://github.com/groovywave/JS_Practice/commit/eaf69bcf87af238e5a61ed71e80207912cf0e133)

Thank you haru-san :)
I removed the ectensions.json .
8fed29a9d2a105a7143052511d56fa2d30a50a26

kenmori commented on src/lesson24/js/main.js:
```diff
@@ -0,0 +1,125 @@
+const password = document.getElementById("password");
+const confirmPassword = document.getElementById("confirmPassword");
+
+function tabAccessControl(target) {
```
名前が微妙です。tabのアクセスをコントロールとは...
おそらくやっていることが多くありすぎるのでそういう名前になるのかなと思いました。切り分けることを検討してください

groovywave replied:
Thank you for your advice :)
I refactored the code.
7a0210284352863682f088a92cb20c63acefc333

kenmori replied:
nits
```js
const focusableElements = [
  ...form.querySelectorAll("a[href], input:not([disabled])"),
];
```
は返ってくるのが既に配列なので ...やる必要ないです

groovywave replied:
Thank you for your  advice ken-san:)
I want to use  array's method indexOf(), so I want to use spread syntax(...).

kenmori commented on src/lesson24/js/main.js:
```diff
@@ -0,0 +1,125 @@
+}
+
+const closeButton = document.getElementById("js-closeButton");
+closeButton.addEventListener("click", () => {
+  closeModal();
+});
```
```suggestion
closeButton.addEventListener("click",  closeModal);
```

groovywave replied:
Thank you ken-san :)
I refactored that.
6608c5e90838c75b073ce2eeb47faeb6d6646062

sae-github commented on package.json:
```diff
@@ -6,15 +6,27 @@
+    "eslint-plugin-tailwindcss": "^3.13.1",
+    "postcss": "^8.4.32",
+    "prettier": "^3.1.1",
+    "prettier-plugin-tailwindcss": "^0.5.9",
```
configファイルに追加する必要がありそうです！
> Then add the plugin to your [Prettier config](https://prettier.io/docs/en/configuration.html):
https://github.com/tailwindlabs/prettier-plugin-tailwindcss

groovywave replied:
Thank you sae-san :)
I added the file.
9ccd1f9d2e58566d2639ee8a257c526063ce3731

sae-github commented on tests/JS_lesson.spec.js:
```diff
@@ -0,0 +1,153 @@
+
+import { test, expect } from "@playwright/test";
+
+test("transition to rule page", async ({ page }) => {
```
IMO
利用規約モーダルが開くかどうかのテストであれば、input要素のテストは別のtestとして書くとわかりやすいと思います！

```
 await page.getByPlaceholder("Enter username").click();
 
 ```

sae-github commented on tests/JS_lesson.spec.js:
```diff
@@ -0,0 +1,153 @@
+  await takeScreenshot("click ruleButton");
+});
+
+test("transition to form page", async ({ page }) => {
+  const takeScreenshot = async (name) => {
+    await page.screenshot({ path: `screenshot/${name}.png` });
+  };
+
+  await page.goto("localhost:3000/");
```
Q. ここのテストはどういった内容のテストですか？👀

groovywave replied:
I just inputed the dev server URL in adress bar.
codegen generated the code.

Uploading lesson24_localhost3.mp4…

sae-github replied:
> test("transition to form page",

のテスト内容を指してます！

sae-github commented on tests/JS_lesson.spec.js:
```diff
@@ -0,0 +1,153 @@
+  await takeScreenshot("click agreeButton");
+});
+
+test("transition to registration page", async ({ page }) => {
+  const takeScreenshot = async (name) => {
```
IMO
> transition to registration page

`registration page`だと登録ページを指すと思います。` registration complete page`とかの方が良さそうです。

groovywave replied:
Thank you sae-san :)
I added the test code.
363541c399f67bb57b2312d6e6a1cc6c47995d66

kenmori commented on .github/workflows/playwright.yml:
```diff
@@ -0,0 +1,27 @@
+        run: npm install
+      - name: Install Playwright Browsers
+        run: npx playwright install --with-deps
+      # - name: Start local server
+        # run: npm run dev &
+      # - name: Wait for localhost
+        # run: npx wait-on http://localhost:3000
```
Must
不要なら消すこと

groovywave replied:
Thank you ken-san :)
I removed that.
cb889710ffbcc2beed09a9f3a8b8de3323af3d6c

kenmori commented on playwright.config.js:
```diff
@@ -0,0 +1,84 @@
+// @ts-check
```
これってts-checkしているのですか？(jsファイルですが)

groovywave replied:
Thank you ken-san :)
I removed the comment .
0d54ca0a9e6f3878d6e7876caaded95556c3029a

kenmori commented on src/lesson24/js/main.js:
```diff
@@ -0,0 +1,118 @@
+import { doc } from "prettier";
```
これは...??使ってなさそう?

groovywave replied:
No, I don't use the object ;<
I removed that.
37d2124dc80f65e4d83ccc88d073b051453b4368

kenmori commented on tailwind.config.js:
```diff
@@ -0,0 +1,13 @@
+/** @type {import('tailwindcss').Config} */
+module.exports = {
+  content: ["./src/**/*.{html,js,jsx,ts,tsx,vue}"],
```
細かいですが、使う拡張子だけにした方がいいです

groovywave replied:
Thank you ken-san :）
I fixed the file.
2768a03f3a7c762459f282a308d72000280f6017

kenmori commented on vite.config.js:
```diff
@@ -1,31 +1,32 @@
-import { resolve } from 'path';
-import { glob } from 'glob';
+import { defineConfig } from "vite";
+// import tailwind from "vite-plugin-tailwind";
```
消したいです

groovywave replied:
Thank you ken-san :）
I removed that.
473890049640c8bb53c14b31b1a77fb985882dc1

sae-github commented on tests/JS_lesson.spec.js:
```diff
@@ -0,0 +1,45 @@
+test("press tab-key and focus on 利用規約", async ({ page }) => {
+  await page.goto("localhost:3000/");
+  await page.getByRole("link", { name: "lesson24" }).click();
+  await page.getByPlaceholder("Enter username").click();
+  await page.getByPlaceholder("Enter username").press("Tab");
```
IMO
label要素が存在している場合は、[page.getByLabel()](https://playwright.dev/docs/api/class-page#page-get-by-label)で値を取るといいと思います!

> WHEN TO USE PLACEHOLDER LOCATORS
Use this locator when locating form elements that do not have labels but do have placeholder texts

ref: https://playwright.dev/docs/locators#locate-by-placeholder

groovywave replied:
Thank you sae-san:）
I learned about it.
b77ae65bf230b413365deadb89c1220c37b8937d

sae-github commented on tests/JS_lesson.spec.js:
```diff
@@ -0,0 +1,45 @@
+});
+
+test("press Enter on the link to rule-page", async ({ page }) => {
+  await page.goto("localhost:3000/");
+  await page.getByRole("link", { name: "lesson24" }).click();
```
IMO
毎回同じ処理を実行する必要がある場合はbeforeを使用するといいと思います！


> In order to avoid repetition for a particular part of your test you can use [before and after hooks](https://playwright.dev/docs/api/class-test).
https://playwright.dev/docs/best-practices#make-tests-as-isolated-as-possible

groovywave replied:
Thank you sae-san :)
I learned a lot of things!!
fc2e4d8dab41234ce87901bf28629042821678ba

sae-github commented on tests/JS_lesson.spec.js:
```diff
@@ -0,0 +1,45 @@
+test("press Enter on the link to rule-page", async ({ page }) => {
+  await page.goto("localhost:3000/");
+  await page.getByRole("link", { name: "lesson24" }).click();
+  await page.getByPlaceholder("Enter password again").click();
+  await page.getByPlaceholder("Enter password again").press("Tab");
```
IMO
```js
await page.getByPlaceholder("Enter password again").click();
  await page.getByPlaceholder("Enter password again").press("Tab");
  ```

このテストでは利用規約モーダルが開くことをチェックすることが目的だと思うので
上記の部分は不要だと思いました！

groovywave replied:
Thank you sae-san:)
I added beforeEach() in my test code and removed that.

sae-github commented on tests/JS_lesson.spec.js:
```diff
@@ -0,0 +1,45 @@
+  await page.getByRole("link", { name: "lesson24" }).click();
+  await page.getByPlaceholder("Enter password again").click();
+  await page.getByPlaceholder("Enter password again").press("Tab");
+  await page.getByRole("link", { name: "利用規約" }).press("Enter");
+});
```
IMO
利用規約を開いてモーダルがオープンになったテストを書いた方が目的に合っているかと思います。
例えば、モーダル開閉時にJSで付け替えている[class名](https://playwright.dev/docs/api/class-locatorassertions#locator-assertions-to-have-class)や、[aria-hiddenの値](https://playwright.dev/docs/api/class-locatorassertions#locator-assertions-to-have-attribute)があればそれらの有無で確認するのがいいと思います。

groovywave replied:
Thank you sae-san:)
I added like the code.
ce4888933569aced54078d8a26a65a819ced80b1

sae-github reviewed (approved):
主にテストの部分にいくつかコメントしましたが全てIMOです！

sae-github commented on src/lesson24/js/main.js:
```diff
@@ -0,0 +1,116 @@
+const focusableElements = [
+  ...form.querySelectorAll("a[href], input:not([disabled])"),
+];
+console.log(focusableElements);
```
不要なconsole.logが混じっているようです！

groovywave replied:
I'm sorry...
e39d90d78573cc495351641ba6bd878186dc83cd

haru-programming commented on src/lesson24/js/main.js:
```diff
@@ -0,0 +1,116 @@
+const modal = document.getElementById("js-modal");
+const modalBody = document.getElementById("js-modalBody");
+
+function keydownToScrollModal(event) {
+  if (modal.classList.contains("hidden")) return;
+  switch (event.key) {
+    case "ArrowUp":
+      return modalBody.scrollBy(0, -16);
+    case "ArrowDown":
+      return modalBody.scrollBy(0, 16);
+  }
+}
```
Q.
こちらは必要でしょうか？
上矢印や下矢印を押すとスクロールするのはデフォルトの挙動だと認識していました👀
その下のlinkToRule.addEventListener 〜でmodalBody.focus();もされているので、この関数はなくても動くのかなと思ったのですが、何か意図があれば教えてください🙏

groovywave replied:
I don't understand why, without this code, the modal can't scroll when the ArrowDown key is pressed:<
With this code, the modal can scroll when the ArrowDown key is pressed.

haru-programming replied:
@haru-programming 
forkして、該当箇所をコメントアウトしてみたのですが、モーダル内を上下の矢印でスクロールできないでしょうか？
[StackBlitz fork ver.](https://stackblitz.com/edit/stackblitz-starters-mnqxdl?file=src%2Flesson24%2Fjs%2Fmain.js)

groovywave replied:
Thank you haru-san :）
I removed that.
0a334924b4a424ba93001439b67f12f7699da023

haru-programming replied:
Q. 追加されていたこちらはどのような意味でしょうか？
removeしているだけでなく、コードを追加されているようなのでこのような場合はご説明いただけると嬉しいです🙇‍♀️
このコードがあると、上下キーでスクロールした後にTabキーでボタンを選択できなくなってしまうと思いました。
```js
function tabKeyIsDisabled(event) {
  if (modal.classList.contains('hidden')) return;
  switch (event.key) {
    case 'Tab':
      return event.preventDefault();
  }
}
modalBody.addEventListener('keydown', tabKeyIsDisabled);
```

groovywave replied:
I apologize for not leaving a comment;< 
The Tab key was disabled because pressing it caused a jump directly to the button without scrolling.

haru-programming replied:
すべて文章をよんでいただけるようにという意図があったのですね👀
今のままだとボタンが押せないので、
IntersectionOvserverが発火したときにはtabキーを有効にするなど工夫が必要そうですね👀

groovywave replied:
Thank you haru-san:)
I added eventLisner on modal.
b78bbbecef2a6d32a7be5e2ce6b011a8cba8e8fa

haru-programming replied:
ごめんなさい、私が言ったのですが、よくよく考えたらスクリーンリーダーは矢印キーで上下させて見えているところを読むわけではないので（順番を指示していない場合は、画面の左上から右下に向かって順に自動的に読んでいく）、この方法は良くないと思いました。

・「上下キーでスクロールして最後までいかないとボタンが押せない」仕様になっていると思うのですが、それはスクリーンリーダーを使用している人には伝わらない。
・ボタンが「利用規約を読み同意しました」となっているため、読んだ上での同意を求めているのでTabキーで最初からfocusが当たる状態で良いと思いました。

haru-programming replied:
仕様外なのでマージでも大丈夫ですー !

haru-programming reviewed (commented):
IMO
同意するにチェックを入れたあと、Tabキーで送信ボタンに飛べないのが気になりました。
キーボード操作は使用外なのですが、他の部分は実装されていたので書かせていただきました🙋‍♀️（スルーでも仕様上は問題ないです🙏）

groovywave commented on .gitignore:
```diff
@@ -22,3 +22,11 @@ dist-ssr
+/test-results/
+/playwright-report/
+/blob-report/
+/playwright/.cache/
```
Thank you sae-san :）
I removed that.
2ffa61e430b814c3b67227ede9ef23fd4d7cc13f

groovywave commented on tests/JS_lesson.spec.js:
```diff
@@ -0,0 +1,45 @@
+
+test("press tab-key and focus on 利用規約", async ({ page }) => {
+  await page.goto("localhost:3000/");
+  await page.getByRole("link", { name: "lesson24" }).click();
```
Thank you sae-san :）
I merged these.
798749b7190eb8bee9de02d0eeb77e8140f5cf56

groovywave commented on src/lesson24/index.html:
```diff
@@ -0,0 +1,421 @@
+        <div class="relative mb-5">
+          <label for="email" class="mb-1 block text-gray-700">Email</label>
+          <input
+            type="text"
```
Thank you sae-san :)
I chenged that.
7c35566a9908f01cdff9f9b2f3c9cc8885cfb615

groovywave commented:
> IMO 同意するにチェックを入れたあと、Tabキーで送信ボタンに飛べないのが気になりました。 キーボード操作は仕様外なのですが、他の部分は実装されていたので書かせていただきました🙋‍♀️

Thank you haru-san:)
I fixed that.
99974ebdd803f089dce5dfa8eb0fd0ec86a448f3
4972b943fd2a899f2aa474248c5a345df2276474

sae-github commented on tests/JS_lesson.spec.js:
```diff
@@ -0,0 +1,35 @@
+  await expect(page.getByRole("link", { name: "利用規約" })).toBeFocused();
+});
+
+test("press tab-key and focus on 利用規約", async () => {});
```
IMO
不要であれば消しましょう！

groovywave replied:
Thank you sae-san :)
I added a test code.
cb7a0657f54820e8015748c7692f7070acfddae8

sae-github commented on tests/JS_lesson.spec.js:
```diff
@@ -0,0 +1,35 @@
+
+test("press tab-key and focus on 利用規約", async () => {});
+
+test("press Enter on the link to rule-page", async ({ page }) => {
```
nits
モーダルが表示されているかどうかのテストだと思うので、それがわかるtestの名前にする方が良さそうです！
めっちゃ細かくてすみません。

groovywave replied:
Thank you sae-san :)
I fixed that.
d03a01afa3ca6507a7cd70a63c2e6ae818ae1e00

sae-github commented on tests/JS_lesson.spec.js:
```diff
@@ -0,0 +1,35 @@
+test("Click the 'Agree' button to go to the form page", async ({ page }) => {
+  await page.getByRole("link", { name: "利用規約" }).press("Enter");
+  await page.getByTestId("js-agreeButton").scrollIntoViewIfNeeded();
+  await page.getByLabel("利用規約", { exact: true }).press("ArrowDown");
```
Q. 
```js
 await page.getByLabel("利用規約", { exact: true }).press("ArrowDown");
```
これは必要ですか？
モーダルが開いている状態で対象の要素をArrowDownするケースはないと思った次第です。

groovywave replied:
Thank you sae-san:)
I removed the code.
c8321170403e3f54a07c1c41adb9726f07a95cf6

sae-github commented on tests/JS_lesson.spec.js:
```diff
@@ -0,0 +1,35 @@
+  await expect(page.getByTestId("js-modalBody")).toBeVisible();
+});
+
+test("Click the 'Agree' button to go to the form page", async ({ page }) => {
```
IMO
テストしたいこと("Click the 'Agree' button to go to the form page")と内容が合っていないと思います。
クリックしたあとに、モーダルが閉じていることを確認する方が、それを証明することになるかと思いました。

groovywave replied:
Thank you sae-san:)
I added the code.
abe152ad2e04174474aa41652ce3431e50ada548

sae-github commented on tests/JS_lesson.spec.js:
```diff
@@ -0,0 +1,35 @@
+  await page.getByText("利用規約を読み、同意しました", { exact: true }).click();
+});
+
+test("Click the 'Submit' button to go to the registration complete page", async ({
+  page,
+}) => {
+  await page.getByRole("link", { name: "利用規約" }).press("Enter");
+  await page.getByTestId("js-agreeButton").scrollIntoViewIfNeeded();
+  await page.getByLabel("利用規約", { exact: true }).press("ArrowDown");
+  await page.getByText("利用規約を読み、同意しました", { exact: true }).click();
+  await page.getByRole("button", { name: "Submit" }).click();
+});
```
IMO
上記のコメントと同様で、テストしたいことと内容が合っていないと思います。
ここでテストしたいのは完了ページに遷移したことなので、そのことが確認できるテストを書くといいと思います！
一例ですが、下記のようにページのheadingタグで確認するのも1つの方法かと思います
```js
  await expect(page.getByRole('heading', { name: 'Your Registration Has Been Completed!' })).toBeVisible();
```

groovywave replied:
Thank you sae-san:)
I added the test code.
c8321170403e3f54a07c1c41adb9726f07a95cf6
I learned a lot of things!!

groovywave commented on tests/JS_lesson.spec.js:
```diff
@@ -0,0 +1,35 @@
+  await page.getByLabel("Password", { exact: true }).press("Tab");
+  await page.getByLabel("Confirm Password").press("Tab");
+  await expect(page.getByRole("link", { name: "利用規約" })).toBeFocused();
+});
```
Thank you sae-san :)
I fixed that.

haru-programming commented:
Q. この辺りのコードって必要そうでしょうか？
先ほどコメントさせていただいた箇所と一緒に、コメントアウトしてみました。
https://github.com/groovywave/JS_Practice/pull/63#discussion_r1462419590
[StackBlitz fork ver.2](https://stackblitz.com/edit/stackblitz-starters-l8ssyf?file=src%2Flesson24%2Fjs%2Fmain.js)

sae-github commented on tests/JS_lesson.spec.js:
```diff
@@ -0,0 +1,41 @@
+  await page.getByLabel('利用規約', { exact: true }).press('ArrowDown');
+  await page.getByText('利用規約を読み、同意しました', { exact: true }).click();
+  await expect(page.getByTestId('js-modalBody')).toBeHidden();
+  await expect(page.getByLabel('Username')).toBeVisible();
+});
+
```
IMO
モーダルを閉じていること確認したいのであれば、`await expect(page.getByTestId('js-modalBody')).toBeHidden();`で満たしているかと思います。
Usernameはモーダルが開いていてもvisibleな要素なのでモーダルを閉じていることのテストにはならないと思います


> Visible
Element is considered visible when it has non-empty bounding box and does not have visibility:hidden computed style.

> Note that according to this definition:

> Elements of zero size are not considered visible.
Elements with display:none are not considered visible.
Elements with opacity:0 are considered visible.
ref: https://playwright.dev/docs/actionability#visible

groovywave replied:
Thank you sae-san:)
I fixed that.
6fe1a4c7ec449e948ce2d2d060facf1b57f95788

groovywave commented:
> Q. この辺りのコードは不要なのかなと思ったのですが、どのような意味があるのかご教授いただきたいです🙇‍♀️
> 
> ```js
> let focusableElements;
> let currentFocusedItemIndex;
> function getFocusableElements(element) {
>   focusableElements = [
>     ...element.querySelectorAll(
>       'a[href], input:not([disabled]),button:not([disabled])'
>     )
>   ];
> }
> function addPressTabEventOnElement(event) {
>   if (event.key !== 'Tab') return;
>   event.preventDefault();
>   currentFocusedItemIndex = focusableElements.indexOf(document.activeElement);
>   focusableElements[
>     (currentFocusedItemIndex + 1) % focusableElements.length
>   ].focus();
> }
> getFocusableElements(form);
> form.addEventListener('keydown', addPressTabEventOnElement);
> ```
> 
> 先ほどコメントさせていただいた箇所と一緒に、コメントアウトしてみました。 [#63 (comment)](https://github.com/groovywave/JS_Practice/pull/63#discussion_r1462419590) [StackBlitz fork ver.2](https://stackblitz.com/edit/stackblitz-starters-l8ssyf?file=src%2Flesson24%2Fjs%2Fmain.js)

With this code, pressing the Tab key will focus only on the target element.(won't focus adress bar.)

haru-programming reviewed (approved):
仕様外のコメント続いているため、一旦Approveさせていただきます🙇‍♀️

haru-programming commented:
以下、もし、お手隙の時がありましたら教えてください🙇‍♀️

> With this code, pressing the Tab key will focus only on the target element.(won't focus adress bar.)

こちらは、「アドレスバーにfocusが当たらないようにする」という解釈であっていますか？
だとすると、それはなぜでしょうか？
これはブラウザの標準的な動作で、ユーザーがページの様々な要素にアクセスするための基本的なナビゲーション手段の一つかなと思ったのですが👀（私は普段アドレスバーの挙動は制限せずに実装しているので理由が気になりました👀）

groovywave commented:
> 以下、もし、お手隙の時がありましたら教えてください🙇‍♀️
> 
> > With this code, pressing the Tab key will focus only on the target element.(won't focus adress bar.)
> 
> こちらは、「アドレスバーにfocusが当たらないようにする」という解釈であっていますか？ だとすると、それはなぜでしょうか？ これはブラウザの標準的な動作で、ユーザーがページの様々な要素にアクセスするための基本的なナビゲーション手段の一つかなと思ったのですが👀（私は普段アドレスバーの挙動は制限せずに実装しているので理由が気になりました👀）

Thank you for your advice haru-san :)
I thought making the Tab key focus only on the form would make inputting easier for users.

haru-programming commented:
教えていただいてありがとうございます !
アクセシビリティまだまだ弱いので、関連するReferenceがあればぜひ今度共有していただけると嬉しいです !
アドレスバーにフォーカスを当たらないようにするということだと、今のコードだと再現できていないかもです👀
TabkeyIsDisabled()があるとTabキーが動かないので、そちらだけコメントアウトして確認しました。
https://github.com/groovywave/JS_Practice/assets/67426438/5c12843c-0c5b-460d-b55c-659b96f3b1e6

https://stackblitz.com/edit/stackblitz-starters-ea9r1c?file=src%2Flesson24%2Fjs%2Fmain.js

私もアドレスバーにフォーカスを当てないようにしたことがないのですぐにコードが書けないのですが、
改善ブランチなどで別途やるといいかもと思いました。
（個人的にはアドレスバーは制御しないくていい？と思ってます !）

sae-github commented:
横からすみません。
> ブラウザの標準的な動作で、ユーザーがページの様々な要素にアクセスするための基本的なナビゲーション手段の一つかなと思った

私もはるさんと同意見でアドレスバーへキーボードでアクセスする手段は残しておくのがいいと思います。(別の手段が用意されていればそれでもいいかもです)
キーボード操作するユーザー視点で考えると、アドレスバーから別ページに移動するケースが想定されると思います。

また、W3Cの基準にもコンテンツは予測可能な方法で表示されで、キーボードでアクセスできることが望ましいことが書いてありました。
> Functionality is available from a keyboard
> Content appears and operates in predictable ways
https://www.w3.org/WAI/fundamentals/accessibility-principles/#predictable

groovywave commented:
> 以下、もし、お手隙の時がありましたら教えてください🙇‍♀️
> 
> > With this code, pressing the Tab key will focus only on the target element.(won't focus adress bar.)
> 
> こちらは、「アドレスバーにfocusが当たらないようにする」という解釈であっていますか？ だとすると、それはなぜでしょうか？ これはブラウザの標準的な動作で、ユーザーがページの様々な要素にアクセスするための基本的なナビゲーション手段の一つかなと思ったのですが👀（私は普段アドレスバーの挙動は制限せずに実装しているので理由が気になりました👀）

Thank you for your advice:)
I thought making the Tab key focus only on the form would make it easier for users.

groovywave commented:
> > Thank you for your advice haru-san :)
> > I thought making the Tab key focus only on the form would make inputting easier for users.
> 
> 教えていただいてありがとうございます ! アクセシビリティまだまだ弱いので、関連するReferenceがあればぜひ今度共有していただけると嬉しいです ! アドレスバーにフォーカスを当たらないようにするということだと、今のコードだと再現できていないかもです👀 TabkeyIsDisabled()があるとTabキーが動かないので、そちらだけコメントアウトして確認しました。 https://github.com/groovywave/JS_Practice/assets/67426438/5c12843c-0c5b-460d-b55c-659b96f3b1e6
> 
> https://stackblitz.com/edit/stackblitz-starters-ea9r1c?file=src%2Flesson24%2Fjs%2Fmain.js
> 
> 私もアドレスバーにフォーカスを当てないようにしたことがないのですぐにコードが書けないのですが、 改善ブランチなどで別途やるといいかもと思いました。 （個人的にはアドレスバーは制御しないくていい？と思ってます !）

Thank you haru-san :)
I learned aout that.
77a509dc0beaa6e2f8f21d9deadd5050b09125ac

groovywave commented:
> 横からすみません。
> 
> > ブラウザの標準的な動作で、ユーザーがページの様々な要素にアクセスするための基本的なナビゲーション手段の一つかなと思った
> 
> 私もはるさんと同意見でアドレスバーへキーボードでアクセスする手段は基本的に残しておくのがいいと思います。 ターゲットとするユーザーや目的によっては、ありなのかもですが代替え手段は必要だと個人的には思います！
> 
> キーボード操作するユーザー視点で考えると、アドレスバーから別ページに移動するケースが想定されると思います。
> 
> また、W3Cの基準にもコンテンツは予測可能な方法で表示されで、キーボードでアクセスできることが望ましいことが書いてありましたので共有します👀
> 
> > Functionality is available from a keyboard
> > Content appears and operates in predictable ways
> > https://www.w3.org/WAI/fundamentals/accessibility-principles/#predictable
> 
> 全くもって仕様外の議論なので対応はお任せします！

Thank you sae-san:)
Thank you for the article:)
77a509dc0beaa6e2f8f21d9deadd5050b09125ac
